src/app-bundles/districts-and-basins-bundle.js

```diff
@@ -90,32 +92,110 @@ export default createRestBundle( {
       }
     ),
-  
-    doSetSelectedDistrict: id => ({
-      type: DISTRICTS_AND_BASINS_SET_DISTRICT_ID,
-      payload: {
-        district_office_id: id,
-        basin_location_id: null
-      },
-    }),
+
+    doSetSelectedDistrict: ( id ) => ( {
+        type: DISTRICTS_AND_BASINS_SET_DISTRICT_ID,
+        payload: {
+          district_id: id,
+          basin_id: "",
+        }
+    } ),
+
+    doUpdateDistrictBasinMapQuery: ( requiresDistrictAndBasin ) => ( { store, dispatch } ) => {
+      const districtId = store.selectSelectedDistrict();
+      const basinId = store.selectSelectedBasin();
+      const newQuery = Object.assign( {}, store.selectQueryObject() );
+      delete newQuery.districtId;
+      delete newQuery.basinId;
+
+      if( !districtId && !basinId ) {
+        store.doUpdateQuery( newQuery );
+        return;
+      }
+
+      const data = store.selectDistrictsAndBasinsData();
+      const district = data[ districtId ];
+      let basin;
+
+      if( district ) newQuery.districtId = district.district_id;
+      let isDistrictOnly = true;
+
+      if( basinId ) {
+        if( district ) basin = district.basins.find( item => item.basin_id === basinId );
+        else {
+          Object.values( data ).some( district => {
+            basin = district.basins.find( basin => basin.basin_id === basinId );
+            return basin;
+          } );
+        }
+
+        // If we have a basin, use the basin's lat/lon
+        if( basin ) {
+          newQuery.basinId = basin.basin_id;
+          newQuery.lat = basin.latitude;
+          newQuery.lon = basin.longitude;
+          newQuery.zoom = 8;
+          isDistrictOnly = false;
+        }
+      }
+
+      // If the flag was passed to require both district and basin, only continue if both are present.
+      if( requiresDistrictAndBasin && isDistrictOnly ) return;
+
+      // If we only have a district, use average lat/lon from the child basins
+      if( isDistrictOnly && district.basins.length > 0 ) {
+        newQuery.districtId = districtId;
+        newQuery.lat = district.basins.reduce( ( acc, b) => acc + b.latitude, 0 ) / district.basins.length;
+        newQuery.lon = district.basins.reduce( ( acc, b) => acc + b.longitude, 0 ) / district.basins.length;
+        newQuery.zoom = 7;
+      }
+
+      store.doUpdateUrl( { pathname: RoutePaths.Map, query: newQuery } );
+    },
 
     doSetSelectedBasin: id => ( {
       type: DISTRICTS_AND_BASINS_SET_BASIN_ID,
       payload: {
-        basin_location_id: id,
+        basin_id: id,
       },
     }),
 
-    selectSelectedDistrict: state => state[name].district_office_id,
+    selectSelectedDistrict: state => state[name].district_id,
+
+    selectSelectedBasin: state => state[name].basin_id,
+
+    selectDistrictBasinQueryChecked: state => state[name].query_checked,
+
+    // To set selected district and basin from URL, if state values have not been set yet.
+    reactShouldSetDistrictBasinFromQuery: createSelector(
```

Comment:
This reactor will only update the `district_id` and `basin_id` in the redux store once, but after that it won't work. For example if the user is on the map page and selects "Baltimore District" then "Detroit District", then clicks the back button, the map will go back to the "Baltimore District", but the district dropdown will still say "Detroit District". In general I don't think it's a good idea to store the district and basin id in both redux and the url, this might lead to more bugs down the road. It might be a better idea to create separate district and basin dropdowns for the map page which aren't connected to redux, and are updated only through the url